presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java

```diff
@@ -1755,20 +1764,86 @@ private static Number toNumber(Object value)
         return (maxRows > 0) ? new LengthLimitedIterator<>(rowsIterator, maxRows) : rowsIterator;
     }
 
+    private static class AsyncIterator<T>
+            implements Iterator<T>
+    {
+        private static final int MAX_QUEUED_ROWS = 50_000;
+        private static final ExecutorService executorService = newCachedThreadPool(daemonThreadsNamed("Presto JDBC worker-%d"));
+
+        private final BlockingQueue<T> rowQueue = new LinkedBlockingQueue<>(MAX_QUEUED_ROWS);
+        Iterator<T> dataIterator;
+        CompletableFuture<Void> future;
+
+        public AsyncIterator(Iterator<T> dataIterator)
+        {
+            this.dataIterator = dataIterator;
+            future = CompletableFuture.supplyAsync(() -> {
+                while (dataIterator.hasNext()) {
+                    try {
+                        rowQueue.put(dataIterator.next());
+                    }
+                    catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(new SQLException("ResultSet thread was interrupted", e));
+                    }
+                }
+                return null;
+            }, executorService);
+        }
+
+        public void cancel()
+        {
+            future.cancel(true);
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            if (!rowQueue.isEmpty()) {
+                return true;
+            }
+            while (rowQueue.isEmpty() && !future.isDone()) {
```

Comment:
This will busy loop, consuming 100% CPU for the calling thread. The original code works because `StatementClient.advance()`, which is used in the `ResultsPageIterator.hasNext()` method, will block.

The difficulty here is we need to need to wait on two conditions: data available or error. We can handle this by adding a dummy value into the queue that signals an error. Then we can simply wait on the queue.

Reply:
The 2nd condition is not just error but is also for checking the possibility of data availability shortly. We cannot get rid of 2nd condition completely because somewhere we will have to wait for the future to be completed before deciding on hasNext() output. Handling the error using a dummy value does not sound to be a clean solution.
We have not observed any problem because of this in our clusters as this condition will not be true for more time.